### ISSUE TITLE
Redirect requests for Whitehall attachments associated with unpublished/withdrawn documents

### ISSUE THREAD

**Issue by floehopper** [alphagov/asset-manager#467]
In Whitehall the `AttachmentsController#show` action will redirect requests for attachments associated with unpublished/withdrawn documents. See https://github.com/alphagov/whitehall/blob/68f93f28ac1d40fedc9b49884ec30c599252cb71/app/controllers/attachments_controller.rb#L43-L44. We'll need to reproduce this for Whitehall assets served from Asset Manager and the redirect URL will need to be kept up-to-date like the draft status.

Note that we'll need to make changes so this working on both an ongoing basis and then doing it for all the existing assets. We might want to split this into two issues.

**Comment by floehopper** [comment]
I'm going to update this issue so that it only includes the redirects for attachments on unpublished/withdrawn documents, and open a new issue to include the redirect for attachments replaced by other attachments.

**Comment by floehopper** [comment]
I've created https://github.com/alphagov/asset-manager/issues/487 to cater for the aspects of this issue which I've just removed.

**Comment by floehopper** [comment]
The Whitehall aspects of this were implemented in https://github.com/alphagov/whitehall/pull/3795 which has just been merged. The Asset Manager aspects of this were implemented in https://github.com/alphagov/asset-manager/pull/481 which was merged yesterday.

**Comment by floehopper** [comment]
https://github.com/alphagov/whitehall/pull/3795 was deployed to integration in `release_13284` just now.

**Comment by floehopper** [comment]
I fixed a bug in https://github.com/alphagov/whitehall/pull/3795 in https://github.com/alphagov/whitehall/pull/3799 and tested it in integration yesterday afternoon at about 5pm.

**Comment by floehopper** [comment]
## Testing in integration

### After attachment added to draft document

* Document in admin: https://whitehall-admin.integration.publishing.service.gov.uk/government/admin/publications/801003
* Attachment in admin: https://whitehall-admin.integration.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/672639/2018-01-14_netLec7.pdf
* Document in preview: https://draft-origin.integration.publishing.service.gov.uk/government/publications/jm-testing-21-feb-2018-v2
* Attachment in preview: https://draft-origin.integration.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/672639/2018-01-14_netLec7.pdf

```
irb> wa = WhitehallAsset.find_by(legacy_url_path: '/government/uploads/system/uploads/attachment_data/file/672639/2018-01-14_netLec7.pdf')
=> #<WhitehallAsset _id: 5a8da5235d7f093e6eeb66fe, deleted_at(deleted_at): nil, created_at: "2018-02-21T16:58:11.303Z", updated_at: "2018-02-21T16:58:34.302Z", state: "uploaded", filename_history: [], uuid: "93bb2227-7e3c-4e0c-a939-be8fe605b2ab", draft: true, redirect_url: "", etag: "5a8da523-5c414", last_modified: "2018-02-21T16:58:11.389Z", md5_hexdigest: "3a1be7a717309631fcd2cb03cd8c0048", size: 377876, access_limited: [], file: "2018-01-14_netLec7.pdf", _type: "WhitehallAsset", legacy_url_path: "/government/uploads/system/uploads/attachment_data/file/672639/2018-01-14_netLec7.pdf", legacy_etag: nil, legacy_last_modified: nil>

irb> wa.state
=> "uploaded"

irb> wa.draft?
=> true

irb> wa.redirect_url
=> ""
```

* Attachment successfully sent to Asset Manager, virus-scanned & uploaded to S3
* Attachment correctly marked as draft
* Attachment redirect URL correctly set to empty string
  * TODO: why not `nil`

### After document force-published

* Document on website: https://www-origin.integration.publishing.service.gov.uk/government/publications/jm-testing-21-feb-2018-v2
* Attachment on website: https://www-origin.integration.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/672639/2018-01-14_netLec7.pdf

```
irb> wa.reload
=> #<WhitehallAsset _id: 5a8da5235d7f093e6eeb66fe, deleted_at(deleted_at): nil, created_at: "2018-02-21T16:58:11.303Z", updated_at: "2018-02-21T17:00:00.321Z", state: "uploaded", filename_history: [], uuid: "93bb2227-7e3c-4e0c-a939-be8fe605b2ab", draft: false, redirect_url: "", etag: "5a8da523-5c414", last_modified: "2018-02-21T16:58:11.389Z", md5_hexdigest: "3a1be7a717309631fcd2cb03cd8c0048", size: 377876, access_limited: [], file: "2018-01-14_netLec7.pdf", _type: "WhitehallAsset", legacy_url_path: "/government/uploads/system/uploads/attachment_data/file/672639/2018-01-14_netLec7.pdf", legacy_etag: nil, legacy_last_modified: nil>

irb> wa.draft?
=> false

irb> wa.redirect_url
=> ""
```

* Attachment correctly marked as not draft
* Attachment redirect URL correctly set to empty string
  * TODO: why not `nil`

### After document unpublished

```
irb> wa.reload
=> #<WhitehallAsset _id: 5a8da5235d7f093e6eeb66fe, deleted_at(deleted_at): nil, created_at: "2018-02-21T16:58:11.303Z", updated_at: "2018-02-21T17:11:26.977Z", state: "uploaded", filename_history: [], uuid: "93bb2227-7e3c-4e0c-a939-be8fe605b2ab", draft: false, redirect_url: "/government/publications/jm-testing-21-feb-2018-v2", etag: "5a8da523-5c414", last_modified: "2018-02-21T16:58:11.389Z", md5_hexdigest: "3a1be7a717309631fcd2cb03cd8c0048", size: 377876, access_limited: [], file: "2018-01-14_netLec7.pdf", _type: "WhitehallAsset", legacy_url_path: "/government/uploads/system/uploads/attachment_data/file/672639/2018-01-14_netLec7.pdf", legacy_etag: nil, legacy_last_modified: nil>

irb> wa.draft?
=> false

irb> wa.redirect_url
=> "/government/publications/jm-testing-21-feb-2018-v2"
```

* Attachment correctly marked as not draft
* Attachment redirect URL correctly set parent document URL path
  * TODO: might need to include host, especially when attachment served from Asset Manager

### After document force-published (again)

```
irb> wa.reload
=> #<WhitehallAsset _id: 5a8da5235d7f093e6eeb66fe, deleted_at(deleted_at): nil, created_at: "2018-02-21T16:58:11.303Z", updated_at: "2018-02-22T10:33:19.300Z", state: "uploaded", filename_history: [], uuid: "93bb2227-7e3c-4e0c-a939-be8fe605b2ab", draft: false, redirect_url: "", etag: "5a8da523-5c414", last_modified: "2018-02-21T16:58:11.389Z", md5_hexdigest: "3a1be7a717309631fcd2cb03cd8c0048", size: 377876, access_limited: [], file: "2018-01-14_netLec7.pdf", _type: "WhitehallAsset", legacy_url_path: "/government/uploads/system/uploads/attachment_data/file/672639/2018-01-14_netLec7.pdf", legacy_etag: nil, legacy_last_modified: nil>

irb> wa.draft?
=> false

irb> wa.redirect_url
=> ""
```

* Attachment correctly marked as not draft
* Attachment redirect URL correctly set to empty string
  * TODO: why not `nil`

### After document withdrawn

```
irb> wa.reload
=> #<WhitehallAsset _id: 5a8da5235d7f093e6eeb66fe, deleted_at(deleted_at): nil, created_at: "2018-02-21T16:58:11.303Z", updated_at: "2018-02-22T10:33:19.300Z", state: "uploaded", filename_history: [], uuid: "93bb2227-7e3c-4e0c-a939-be8fe605b2ab", draft: false, redirect_url: "", etag: "5a8da523-5c414", last_modified: "2018-02-21T16:58:11.389Z", md5_hexdigest: "3a1be7a717309631fcd2cb03cd8c0048", size: 377876, access_limited: [], file: "2018-01-14_netLec7.pdf", _type: "WhitehallAsset", legacy_url_path: "/government/uploads/system/uploads/attachment_data/file/672639/2018-01-14_netLec7.pdf", legacy_etag: nil, legacy_last_modified: nil>

irb> wa.draft?
=> false

irb> wa.redirect_url
=> ""
```

* Attachment correctly marked as not draft
* Attachment redirect URL correctly set to empty string
  * TODO: why not `nil`

### After document unwithdrawn

```
irb> wa.reload
=> #<WhitehallAsset _id: 5a8da5235d7f093e6eeb66fe, deleted_at(deleted_at): nil, created_at: "2018-02-21T16:58:11.303Z", updated_at: "2018-02-22T10:33:19.300Z", state: "uploaded", filename_history: [], uuid: "93bb2227-7e3c-4e0c-a939-be8fe605b2ab", draft: false, redirect_url: "", etag: "5a8da523-5c414", last_modified: "2018-02-21T16:58:11.389Z", md5_hexdigest: "3a1be7a717309631fcd2cb03cd8c0048", size: 377876, access_limited: [], file: "2018-01-14_netLec7.pdf", _type: "WhitehallAsset", legacy_url_path: "/government/uploads/system/uploads/attachment_data/file/672639/2018-01-14_netLec7.pdf", legacy_etag: nil, legacy_last_modified: nil>

irb> wa.draft?
=> false

irb> wa.redirect_url
=> ""
```

* Attachment correctly marked as not draft
* Attachment redirect URL correctly set to empty string

## Things to investigate

* Why is Asset#redirect_url set to `""` vs `nil` in Asset Manager?
* Was redirect_url always set as a path without a host?
* Will redirect_url need to include host when attachment served from Asset Manager?

**Comment by floehopper** [comment]
> * Why is Asset#redirect_url set to `""` vs `nil` in Asset Manager?

This seems to be caused by the underlying use of `RestClient` within `gds-api-adapters`. I've implemented a fix in https://github.com/alphagov/asset-manager/pull/488.

**Comment by floehopper** [comment]
https://github.com/alphagov/asset-manager/pull/488 has been merged and deployed to all environments.

**Comment by floehopper** [comment]
To Do:

* [ ] Test this behaviour in staging/production
* [ ] Was `redirect_url` always set as a path without a host?
* [ ] Will `redirect_url` need to include host when attachment served from Asset Manager?

**Comment by chrisroos** [comment]
I'm going to test this behaviour now, and work whether we need to do anything about the second and third bullet points above.

**Comment by chrisroos** [comment]
I've opened https://github.com/alphagov/whitehall/pull/3841 to set the `redirect_url` to be a URL instead of just the URL path. This is required so that we can redirect from draft-assets or assets-origin to www.gov.uk.

**Comment by chrisroos** [comment]
@chrislo and I have now tested this on staging. We created a consultation, attached a file and force published it. We observed that:

* After withdrawing the consultation we could still view the associated attachment using `www-origin` (Whitehall) and `draft-assets` (Asset Manager).
* After unpublishing the consultation we were redirected to the consultation page when requesting the associated attachment using `www-origin` and `draft-assets`.

This all appears to be working as expected so I'm closing this issue.
